API_DESIGN.md

```diff
@@ -0,0 +1,295 @@
+  string user_id = 2 [
+    (validate.rules).string = {min_len: 1, max_len: 200},
+  ];
+  string user_name = 3;
+}
+
+message IdentiyProviderLink {
+  string identity_provider_id = 1;
+  string identity_provider_name = 2;
+  string user_id = 3;
+  string user_name = 4;
+} 
+```
+
+#### Operations and Methods
+
+Methods on a resource should be named using the following convention:
+- Create: `Create<resource>`
+- Update: `Update<resource>`
+- Delete: `Delete<resource>`
+- Get: `Get<resource>`
+- List: `List<resource>`
+- Search: `Search<resource>`
+
+Methods on a list of resources should be named using the following convention:
+- Add: `Add<resource>`
+- Remove: `Remove<resource>`
+- Set: `Set<resource>`
+
+## Authentication and Authorization
+
+The API uses OAuth 2 for authorization. There are corresponding middlewares that check the access token for validity and 
+automatically return an error if the token is invalid.
+
+Permissions grated to the user are organization specific and might only be checked based on the queried resource.
+Therefore, the API does not check the permissions itself but relies on the checks of the functions that are called by the API.
+Required permissions need to be documented in the [API documentation](#documentation).
+
+## Pagination
+
+The API uses pagination for listing resources. The client can specify a limit and an offset to retrieve a subset of the resources.
+Additionally, the client can specify sorting options to sort the resources by a specific field.
+
+Most listing methods should provide use the `ListQuery` message to allow the client to specify the limit, offset, and sorting options.
+```protobuf
+
+// ListQuery is a general query object for lists to allow pagination and sorting.
+message ListQuery {
+  uint64 offset = 1;
+  // limit is the maximum amount of objects returned. The default is set to 100
+  // with a maximum of 1000 in the runtime configuration.
+  // If the limit exceeds the maximum configured ZITADEL will throw an error.
+  // If no limit is present the default is taken.
+  uint32 limit = 2;
+  // Asc is the sorting order. If true the list is sorted ascending, if false
+  // the list is sorted descending. The default is descending.
+  bool asc = 3;
+}
+```
+On the corresponding responses the `ListDetails` can be used to return the total count of the resources
+and allow the user to handle their offset and limit accordingly.
+
+
+## Error Handling
+
+The API returns machine-readable errors in the response body. This includes a status code, an error code and possibly some details.
+
+### Status Codes
+
+The API uses status codes to indicate the status of a request. Depending on the protocol used to call the API,
+the status code is returned as an HTTP status code or as a gRPC / connectRPC status code.
+Check the possible status codes https://zitadel.com/docs/apis/statuscodes
+
+### Error Codes
+
+Additionally to the status code, the API returns unique error codes for each type of error.
```

Comment:
Is it not rather `each occurrence of error type` instead of `each type of error`?

Reply:
would `each occurrence of error type` mean that if an error type is returned in multiple places (even in the same function) it would need a unique code?

Reply:
Yes, to keep it unique to find out exactly where the error comes from, you either have to have error types which can occur multiple times, or unique error types which means you need a type for every specific code usage, or am I misunderstanding something?

Reply:
The error type / code needs to provide information to the caller what was wrong / what needs to be changed and not where it occurred. If an error can occur at multiple spots in the code, that's a problem of logging accordingly and maybe provide an explicit id for the occurence / traceid